### PR TITLE
Positive Unlabeled Learning for stepstoreproduce model

### PR DESCRIPTION
**Pull request by chidauri**
Regarding #705

### CODE REVIEW

**Comment by chidauri** [comment]
@marco-c This is the result
48949 bugs have no steps to reproduce
1051 bugs have steps to reproduce
X: (50000, 570310), y: (50000,)
Cross Validation scores:
Accuracy: f0.8746215073731051 (+/- 0.022296351798910312)
Precision: f0.12633390170707107 (+/- 0.012666598107065403)
Recall: f0.8269275282446955 (+/- 0.06671257274087936)
X_train: (1906, 570310), y_train: (1906,)
X_test: (5000, 570310), y_test: (5000,)
Test Set scores:
No confidence threshold - 5000 classified
                   pre       rec       spe        f1       geo       iba       sup

          1       0.12      0.83      0.88      0.21      0.85      0.72        98
          0       1.00      0.88      0.83      0.93      0.85      0.73      4902

avg / total       0.98      0.88      0.83      0.92      0.85      0.73      5000

╒════════════╤═════════════════╤═════════════════╕
│            │   1 (Predicted) │   0 (Predicted) │
╞════════════╪═════════════════╪═════════════════╡
│ 1 (Actual) │              81 │              17 │
├────────────┼─────────────────┼─────────────────┤
│ 0 (Actual) │             596 │            4306 │
╘════════════╧═════════════════╧═════════════════╛

**Comment by chidauri** [comment]
The Baggingclassifier can be applied in this case by changing the functionality of "_parallel_build_estimators" and "get_estimators_indices" , so this is being done is pulearning.py.

File: bugbug/pulearning.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+# License, v. 2.0. If a copy of the MPL was not distributed with this file,
+# You can obtain one at http://mozilla.org/MPL/2.0/.
+
+# Module for Positive Unlabeled Learning
+import numpy as np
+from sklearn.ensemble import bagging
+from sklearn.utils import indices_to_mask
+
+
+def pu_parallel_build_estimators(
```
Comment (marco-c) [inline bugbug/pulearning.py]: Did you write this code, or did you get from somewhere? If the second, we should put a reference to the origin (and check its license to see if we can use it).

File: bugbug/models/stepstoreproduce.py [hunk 2]
```diff
@@ -58,25 +57,20 @@ def __init__(self, lemmatization=False):
...
-                elif bug_data["cf_has_str"] == "yes":
+                if bug_data["cf_has_str"] == "yes":
                     classes[int(bug_data["id"])] = 1
-            elif "stepswanted" in bug_data["keywords"]:
-                classes[int(bug_data["id"])] = 0
-            else:
-                for entry in bug_data["history"]:
-                    for change in entry["changes"]:
-                        if change["removed"].startswith("stepswanted"):
-                            classes[int(bug_data["id"])] = 1
```
Comment (marco-c) [inline bugbug/models/stepstoreproduce.py]: Instead of modifying everything here, you can just keep it as it was before and then do:

```Python
if int(bug_data["id"]) not in classes:
    classes[int(bug_data["id"])] = 0
```

**Comment by marco-c** [comment]
We need to calculate the metrics we talked about on IRC (that is, if you hide some real positives, is the model able to find them? if you hide some real negatives, is the model able to tell that they are negatives?)

One possible option (maybe easier? Not sure) is to make the test set only contain true positives and negatives.

File: bugbug/pulearning.py (hunk 1 above)
Comment (chidauri) [reply bugbug/pulearning.py]: To be exact https://github.com/roywright/pu_learning/blob/a394b2691818813441eb65270870291fbab194f9/baggingPU.py#L75

Comment (chidauri) [reply bugbug/pulearning.py]: Inspired from [https://github.com/roywright/pu_learning/blob/master/baggingPU.py](url)

Comment (marco-c) [reply bugbug/pulearning.py]: OK, it's BSD. You'll need to copy the notice from the top of that file to here.

**Comment by chidauri** [comment]
@marco-c results after new eval metric
48926 bugs have no labels
23 bugs have no steps to reproduce
1051 bugs have steps to reproduce
X: (50000, 570310), y: (50000,)
Cross Validation scores:
Accuracy: f0.9865714407119537 (+/- 0.0011423387192974253)
Precision: f0.723844058058155 (+/- 0.03863886828612711)
Recall: f0.5855947784519213 (+/- 0.06833560087888832)
X_train: (35000, 570310), y_train: (35000,)
X_test: (323, 570310), y_test: (323,)
Test Set scores:
No confidence threshold - 323 classified
                   pre       rec       spe        f1       geo       iba       sup

          1       0.99      0.56      0.75      0.71      0.65      0.41       315
          0       0.04      0.75      0.56      0.08      0.65      0.43         8

avg / total       0.97      0.56      0.75      0.70      0.65      0.41       323

╒════════════╤═════════════════╤═════════════════╕
│            │   1 (Predicted) │   0 (Predicted) │
╞════════════╪═════════════════╪═════════════════╡
│ 1 (Actual) │             176 │             139 │
├────────────┼─────────────────┼─────────────────┤
│ 0 (Actual) │               2 │               6 │
╘════════════╧═════════════════╧═════════════════╛


Confidence threshold > 0.6 - 315 classified
                   pre       rec       spe        f1       geo       iba       sup

          1       0.99      0.57      0.75      0.73      0.66      0.42       307
          0       0.04      0.75      0.57      0.08      0.66      0.44         8

avg / total       0.96      0.58      0.75      0.71      0.66      0.42       315

╒════════════╤═════════════════╤═════════════════╕
│            │   1 (Predicted) │   0 (Predicted) │
╞════════════╪═════════════════╪═════════════════╡
│ 1 (Actual) │             176 │             131 │
├────────────┼─────────────────┼─────────────────┤
│ 0 (Actual) │               2 │               6 │
╘════════════╧═════════════════╧═════════════════╛


Confidence threshold > 0.7 - 302 classified
                   pre       rec       spe        f1       geo       iba       sup

          1       0.99      0.57      0.75      0.72      0.65      0.42       294
          0       0.05      0.75      0.57      0.09      0.65      0.43         8

avg / total       0.96      0.57      0.75      0.70      0.65      0.42       302

╒════════════╤═════════════════╤═════════════════╕
│            │   1 (Predicted) │   0 (Predicted) │
╞════════════╪═════════════════╪═════════════════╡
│ 1 (Actual) │             167 │             127 │
├────────────┼─────────────────┼─────────────────┤
│ 0 (Actual) │               2 │               6 │
╘════════════╧═════════════════╧═════════════════╛


Confidence threshold > 0.8 - 279 classified
                   pre       rec       spe        f1       geo       iba       sup

          1       0.99      0.56      0.86      0.72      0.69      0.46       272
          0       0.05      0.86      0.56      0.09      0.69      0.49         7

avg / total       0.97      0.57      0.85      0.70      0.69      0.47       279

╒════════════╤═════════════════╤═════════════════╕
│            │   1 (Predicted) │   0 (Predicted) │
╞════════════╪═════════════════╪═════════════════╡
│ 1 (Actual) │             152 │             120 │
├────────────┼─────────────────┼─────────────────┤
│ 0 (Actual) │               1 │               6 │
╘════════════╧═════════════════╧═════════════════╛


Confidence threshold > 0.9 - 239 classified
                   pre       rec       spe        f1       geo       iba       sup

          1       0.99      0.55      0.86      0.71      0.69      0.46       232
          0       0.05      0.86      0.55      0.10      0.69      0.49         7

avg / total       0.96      0.56      0.85      0.69      0.69      0.46       239

╒════════════╤═════════════════╤═════════════════╕
│            │   1 (Predicted) │   0 (Predicted) │
╞════════════╪═════════════════╪═════════════════╡
│ 1 (Actual) │             128 │             104 │
├────────────┼─────────────────┼─────────────────┤
│ 0 (Actual) │               1 │               6 │
╘════════════╧═════════════════╧═════════════════╛

**Comment by chidauri** [comment]
@marco-c **After the latest changes, it performs a lot better now**
48965 bugs have no labels
27 bugs have no steps to reproduce
1008 bugs have steps to reproduce
X: (50000, 575524), y: (50000,)
Cross Validation scores:
Accuracy: f0.8663989202635347 (+/- 0.020209775582136394)
Precision: f0.11253227786380618 (+/- 0.014047676653706294)
Recall: f0.825077081192189 (+/- 0.07402488946446677)
X_train: (35000, 575524), y_train: (35000,)
X_test: (321, 575524), y_test: (321,)
Test Set scores:
No confidence threshold - 321 classified
                   pre       rec       spe        f1       geo       iba       sup

          1       0.99      0.81      0.70      0.89      0.75      0.57       311
          0       0.11      0.70      0.81      0.18      0.75      0.56        10

avg / total       0.96      0.81      0.70      0.87      0.75      0.57       321

╒════════════╤═════════════════╤═════════════════╕
│            │   1 (Predicted) │   0 (Predicted) │
╞════════════╪═════════════════╪═════════════════╡
│ 1 (Actual) │             252 │              59 │
├────────────┼─────────────────┼─────────────────┤
│ 0 (Actual) │               3 │               7 │
╘════════════╧═════════════════╧═════════════════╛


Confidence threshold > 0.6 - 295 classified
                   pre       rec       spe        f1       geo       iba       sup

          1       0.99      0.88      0.57      0.93      0.71      0.52       288
          0       0.10      0.57      0.88      0.17      0.71      0.48         7

avg / total       0.97      0.87      0.58      0.91      0.71      0.51       295

╒════════════╤═════════════════╤═════════════════╕
│            │   1 (Predicted) │   0 (Predicted) │
╞════════════╪═════════════════╪═════════════════╡
│ 1 (Actual) │             252 │              36 │
├────────────┼─────────────────┼─────────────────┤
│ 0 (Actual) │               3 │               4 │
╘════════════╧═════════════════╧═════════════════╛


Confidence threshold > 0.7 - 262 classified
                   pre       rec       spe        f1       geo       iba       sup

          1       0.99      0.92      0.67      0.96      0.78      0.63       256
          0       0.17      0.67      0.92      0.27      0.78      0.60         6

avg / total       0.97      0.92      0.67      0.94      0.78      0.63       262

╒════════════╤═════════════════╤═════════════════╕
│            │   1 (Predicted) │   0 (Predicted) │
╞════════════╪═════════════════╪═════════════════╡
│ 1 (Actual) │             236 │              20 │
├────────────┼─────────────────┼─────────────────┤
│ 0 (Actual) │               2 │               4 │
╘════════════╧═════════════════╧═════════════════╛


Confidence threshold > 0.8 - 231 classified
                   pre       rec       spe        f1       geo       iba       sup

          1       0.99      0.96      0.60      0.98      0.76      0.60       226
          0       0.25      0.60      0.96      0.35      0.76      0.56         5

avg / total       0.97      0.95      0.61      0.96      0.76      0.60       231

╒════════════╤═════════════════╤═════════════════╕
│            │   1 (Predicted) │   0 (Predicted) │
╞════════════╪═════════════════╪═════════════════╡
│ 1 (Actual) │             217 │               9 │
├────────────┼─────────────────┼─────────────────┤
│ 0 (Actual) │               2 │               3 │
╘════════════╧═════════════════╧═════════════════╛


Confidence threshold > 0.9 - 204 classified
                   pre       rec       spe        f1       geo       iba       sup

          1       0.99      0.98      0.50      0.98      0.70      0.51       202
          0       0.17      0.50      0.98      0.25      0.70      0.46         2

avg / total       0.99      0.97      0.50      0.98      0.70      0.51       204

╒════════════╤═════════════════╤═════════════════╕
│            │   1 (Predicted) │   0 (Predicted) │
╞════════════╪═════════════════╪═════════════════╡
│ 1 (Actual) │             197 │               5 │
├────────────┼─────────────────┼─────────────────┤
│ 0 (Actual) │               1 │               1 │
╘════════════╧═════════════════╧═════════════════╛
